### PR TITLE
Rework PosteOpenActions to javafx (custom entry type import)

### PR DESCRIPTION
**Pull request by Siedlerchr**
Convert dialog for importing custom entry types to CheckListView
Rework threading when opening a database

The dialog triggers when you have a custom Entry type in your bib and open it via File -> Open library
Never seen this dialog before...

Before:
CustomEntryTypesImport

Now
grafik

Implement toString() for entryType classes, otherwise the CheckLIstView does not show the name.

I am currently a bit unhappy with the result lists from the dialog. Maybe someone has a more cleaner solution

```
@asdfasdfasdfasdfasdfas{CustomAuthor2016,
  author    = {CustomAuthor},
  date      = {2016},
  timestamp = {2016.09.04},
  year      = {2016},
}
```

@Comment{jabref-entrytype: asdfasdfasdfasdfasdfas: req[author;year/date] opt[]}

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Manually tested changed features in running JabRef
- [ ] Screenshots added in PR description (for bigger UI changes)
- [ ] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

### CODE REVIEW

File: src/main/java/org/jabref/gui/importer/actions/OpenDatabaseAction.java [hunk 1]
```diff
@@ -184,49 +183,56 @@ else if (toRaise != null) {
...
-            if (BackupManager.checkForBackupFile(fileToLoad)) {
-                BackupUIManager.showRestoreBackupDialog(frame.getDialogService(), fileToLoad);
-            }
+    private ResultHelper loadDatabase(Path file, boolean raisePanel) {
+        Path fileToLoad = file.toAbsolutePath();
 
-            ParserResult result;
-            result = OpenDatabase.loadDatabase(fileToLoad.toString(),
-                    Globals.prefs.getImportFormatPreferences(), Globals.getFileUpdateMonitor());
+        frame.getDialogService().notify(Localization.lang("Opening") + ": '" + file + "'");
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/importer/actions/OpenDatabaseAction.java]: this `notify` message is not called on the JavaFX thread

File: src/main/java/org/jabref/gui/importer/actions/OpenDatabaseAction.java [hunk 2]
```diff
@@ -184,49 +183,56 @@ else if (toRaise != null) {
...
+        if (result.getDatabase().isShared()) {
+            try {
+                new SharedDatabaseUIManager(frame).openSharedDatabaseFromParserResult(result);
+            } catch (SQLException | DatabaseNotSupportedException | InvalidDBMSConnectionPropertiesException |
+                     NotASharedDatabaseException e) {
+                result.getDatabaseContext().clearDatabaseFile(); // do not open the original file
+                result.getDatabase().clearSharedDatabaseID();
+                LOGGER.error("Connection error", e);
+
+                frame.getDialogService().showErrorDialogAndWait(Localization.lang("Connection error"),
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/importer/actions/OpenDatabaseAction.java]: same here, not on JavaFX thread. Just rethrow the exception here (after cleaning up) and add `onException` handler to background task above

File: src/main/java/org/jabref/gui/importer/actions/OpenDatabaseAction.java [hunk 3]
```diff
@@ -235,14 +241,14 @@ private BasePanel addNewDatabase(ParserResult result, final Path file, boolean r
 
         if (result.hasWarnings()) {
             JabRefExecutorService.INSTANCE
-                    .execute(() -> ParserResultWarningDialog.showParserResultWarningDialog(result, frame));
+                                          .execute(() -> ParserResultWarningDialog.showParserResultWarningDialog(result, frame));
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/importer/actions/OpenDatabaseAction.java]: why is this on a new thread (which is not the JavaFX one)?

File: src/main/java/org/jabref/gui/importer/actions/OpenDatabaseAction.java [hunk 4]
```diff
@@ -184,49 +183,56 @@ else if (toRaise != null) {
...
 
-            // After adding the database, go through our list and see if
-            // any post open actions need to be done. For instance, checking
-            // if we found new entry types that can be imported, or checking
-            // if the database contents should be modified due to new features
-            // in this version of JabRef:
-            final ParserResult finalReferenceToResult = result;
-            SwingUtilities.invokeLater(() -> OpenDatabaseAction.performPostOpenActions(panel, finalReferenceToResult));
-        }
+        // After adding the database, go through our list and see if
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/importer/actions/OpenDatabaseAction.java]: the comment is not right anymore

File: src/main/java/org/jabref/gui/importer/actions/ResultHelper.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package org.jabref.gui.importer.actions;
+
+import org.jabref.gui.BasePanel;
+import org.jabref.gui.DialogService;
+import org.jabref.logic.importer.ParserResult;
+
+public class ResultHelper {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/importer/actions/ResultHelper.java]: make this an inner (private) class of opendatabaseaction, as this is an implementation detail

File: src/main/java/org/jabref/gui/importer/actions/OpenDatabaseAction.java (hunk 3 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/importer/actions/OpenDatabaseAction.java]: No idea, this was probably already present but came to light due to my formatting of the class. But makes indeed no sense

File: src/main/java/org/jabref/gui/importer/actions/ResultHelper.java (hunk 5 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/importer/actions/ResultHelper.java]: I could get rid of this class and simplified the threading stuff

**Comment by Siedlerchr** [comment]
Reworked to fxml/mvvm and simpliefied code

File: src/main/java/org/jabref/gui/importer/ImportCustomEntryTypesDialogViewModel.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+import org.jabref.Globals;
+import org.jabref.gui.customentrytypes.CustomEntryTypesManager;
+import org.jabref.model.EntryTypes;
+import org.jabref.model.database.BibDatabaseMode;
+import org.jabref.model.entry.CustomEntryType;
+import org.jabref.model.entry.EntryType;
+
+public class ImportCustomEntryTypesDialogViewModel {
+
+    private final ListProperty<EntryType> newTypesProperty;
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/importer/ImportCustomEntryTypesDialogViewModel.java]: There is no need for an `ListProperty` here, an observable list is sufficient (the list property is handy if you want to replace the complete list and need listeners for this).

File: src/main/java/org/jabref/gui/importer/ImportCustomEntryTypesDialogViewModel.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    public ListProperty<EntryType> differentCustomizationsProperty() {
+        return this.differentCustomizationsProperty;
+    }
+
+    public void importCustomEntryTypes(List<EntryType> checkedUnknownEntryTypes, List<EntryType> checkedDifferentEntryTypes) {
+        if (!checkedUnknownEntryTypes.isEmpty()) {
+            checkedUnknownEntryTypes.forEach(type -> EntryTypes.addOrModifyCustomEntryType((CustomEntryType) type, mode));
+            CustomEntryTypesManager.saveCustomEntryTypes(Globals.prefs);
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/importer/ImportCustomEntryTypesDialogViewModel.java]: Please extract the `Globals.prefs` dependency (and even better move the `saveCustomEntryTypes` method to the preferences class)

File: src/main/java/org/jabref/gui/importer/actions/CheckForNewEntryTypesAction.java [hunk 8]
```diff
@@ -39,94 +26,18 @@ public boolean isActionNecessary(ParserResult parserResult) {
     public void performAction(BasePanel panel, ParserResult parserResult) {
         BibDatabaseMode mode = getBibDatabaseModeFromParserResult(parserResult);
 
-        List<EntryType> typesToStore = determineEntryTypesToSave(panel, getListOfUnknownAndUnequalCustomizations(parserResult), mode);
+        ImportCustomEntryTypesDialog dlg = new ImportCustomEntryTypesDialog(mode, getListOfUnknownAndUnequalCustomizations(parserResult));
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/importer/actions/CheckForNewEntryTypesAction.java]: no abbreviations

File: src/main/java/org/jabref/gui/importer/actions/GUIPostOpenAction.java [hunk 9]
```diff
@@ -32,6 +32,7 @@
      *
      * @param panel The BasePanel where the database is shown.
      * @param pr The result of the BIB parse operation.
+     * @param dialogService
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/importer/actions/GUIPostOpenAction.java]: can be removed

File: src/main/java/org/jabref/gui/importer/actions/OpenDatabaseAction.java [hunk 10]
```diff
@@ -49,24 +46,26 @@
...
         this.frame = frame;
+        this.dialogService = frame.getDialogService();
     }
 
     /**
      * Go through the list of post open actions, and perform those that need to be performed.
      *
      * @param panel  The BasePanel where the database is shown.
      * @param result The result of the BIB file parse operation.
+     * @param dialogService
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/importer/actions/OpenDatabaseAction.java]: here too

**Pull request merged**